pyomo/contrib/pynumero/algorithms/solvers/cyipopt_solver.py

```diff
@@ -0,0 +1,299 @@
+            return data
+        data = np.compress(self._hess_lower_mask, self._hess_lag.coo_data())
+        return data
+
+    def intermediate(
+            self,
+            alg_mod,
+            iter_count,
+            obj_value,
+            inf_pr,
+            inf_du,
+            mu,
+            d_norm,
+            regularization_size,
+            alpha_du,
+            alpha_pr,
+            ls_trials
+    ):
+        pass
+
+
+class IpoptSolver(object):
+
+    def __init__(self, nlp, options=None):
+
+        self._nlp = nlp
+
+        self._is_composite = False
+        if hasattr(nlp, 'nblocks'):
+            self._is_composite = True
+
+        self._problem = _CyIpoptProblem(nlp)
+
+        self._options = options
+        if options is not None:
+            assert isinstance(options, dict)
+        else:
+            self._options = dict()
+
+    def solve(self, x0=None, tee=False):
+
+        if not self._is_composite:
+            cyipopt_solver = ipopt.problem(n=self._nlp.nx,
+                                           m=self._nlp.ng,
+                                           problem_obj=self._problem,
+                                           lb=self._nlp.xl(),
+                                           ub=self._nlp.xu(),
+                                           cl=self._nlp.gl(),
+                                           cu=self._nlp.gu()
+                                           )
+        else:
+            xl = self._nlp.xl()
+            xu = self._nlp.xu()
+            gl = self._nlp.gl()
+            gu = self._nlp.gu()
+            nx = int(self._nlp.nx)
+            ng = int(self._nlp.ng)
+            cyipopt_solver = ipopt.problem(n=nx,
+                                           m=ng,
+                                           problem_obj=self._problem,
+                                           lb=xl.flatten(),
+                                           ub=xu.flatten(),
+                                           cl=gl.flatten(),
+                                           cu=gu.flatten()
+                                           )
+        if x0 is None:
+            xstart = self._nlp.x_init()
+            if self._is_composite:
+                xstart = xstart.flatten()
+        else:
+            assert isinstance(x0, np.ndarray)
+            assert x0.size == self._nlp.nx
+            xstart = x0
+
+        # this is needed until NLP hessian takes obj_factor as an input
+        if not self._nlp._future_libraries:
+            cyipopt_solver.addOption('nlp_scaling_method', 'none')
+
+        # add options
+        for k, v in self._options.items():
+            cyipopt_solver.addOption(k, v)
+
+        if not tee:
+            newstdout = redirect_stdout()
+
+        x, info = cyipopt_solver.solve(xstart)
+
+        if not tee:
+            os.dup2(newstdout, 1)
+
+        return x, info
+
+
+def create_model1():
```

Comment:
This method and the ones below it looks like testing functions.  Should they be moved into `pyomo.contrib.pynumero.algorithms.tests`?

Reply:
I agree. As I said before this was more of a experimentation face of CyIpoptSolver. I will write couple of tests and move them to algorithms.tests

Reply:
Done. Added `pyomo.contrib.pynumero.algorithms.tests.test_cyipopt_solver.py`